# Machine Learning Assignment 1

### Imports

In [131]:
import pandas as pd

## Pre-Tasks Analysis of Pulsar Star Dataset

In [132]:
# Reading in CSV and creating first dataframe(df)
df = pd.read_csv('pulsar_star_19202103.csv')
df

Mean of the integrated profile  \
0                          114.859375   
1                           98.406250   
2                          110.359375   
3                          133.960938   
4                          100.109375   
...                               ...   
9995                       122.976562   
9996                       110.257812   
9997                       113.062500   
9998                       123.765625   
9999                       135.523438   

       Standard deviation of the integrated profile  \
0                                         41.452672   
1                                         51.290752   
2                                         42.741544   
3                                         48.455908   
4                                         44.539350   
...                                             ...   
9995                                      48.277108   
9996                                      58.110784   
9997                                      48.387716   
9998                                      41.959706   
9999                                      49.079649   

       Excess kurtosis of the integrated profile  \
0                                       0.188914   
1                                       0.421885   
2                                            NaN   
3                                      -0.289046   
4                                       0.441861   
...                                          ...   
9995                                   -0.027160   
9996                                    0.132559   
9997                                    0.019552   
9998                                   -0.120701   
9999                                    0.004527   

       Skewness of the integrated profile   Mean of the DM-SNR curve  \
0                                0.591629                   2.579431   
1                               -0.158079                   3.683946   
2                                0.538841                   3.418896   
3                                0.093911                   1.377090   
4                                0.350467                   8.473244   
...                                   ...                        ...   
9995                            -0.080098                   5.120401   
9996                            -0.750325                   2.829431   
9997                            -0.362260                   0.352843   
9998                             0.379044                   7.254181   
9999                            -0.419013                   2.815217   

       Standard deviation of the DM-SNR curve  \
0                                   16.024319   
1                                   23.996595   
2                                   19.090287   
3                                   11.289180   
4                                   32.343502   
...                                       ...   
9995                                28.763490   
9996                                18.674093   
9997                                 8.201329   
9998                                35.665727   
9999                                19.911051   

       Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
0                                 9.183242                     105.904593   
1                                 7.378554                      57.771480   
2                                 7.527202                      67.811225   
3                                13.631908                            NaN   
4                                 4.532549                      22.332795   
...                                    ...                            ...   
9995                              5.867339                      34.484277   
9996                              9.247513                      98.431192   
9997                             27.385053                     804.482706   
9998

We have read in our personal dataset(Student Number: 19202103) and we can see there are 10000 rows (detected signals) and 9 columns (8 descriptive features and the target feature).

In [133]:
# Describing the dataframe to identify trends of each column
df.describe()

Mean of the integrated profile  \
count                     10000.000000   
mean                        111.103095   
std                          25.640401   
min                           5.812500   
25%                         101.076172   
50%                         115.128906   
75%                         127.015625   
max                         189.734375   

        Standard deviation of the integrated profile  \
count                                   10000.000000   
mean                                       46.509434   
std                                         6.798077   
min                                        24.772042   
25%                                        42.362222   
50%                                        46.933947   
75%                                        50.970903   
max                                        91.808628   

        Excess kurtosis of the integrated profile  \
count                                 8595.000000   
mean                                     0.477388   
std                                      1.066709   
min                                     -1.738021   
25%                                      0.025565   
50%                                      0.223402   
75%                                      0.473025   
max                                      8.069522   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
count                         10000.000000               10000.000000   
mean                              1.779757                  12.671227   
std                               6.224056                  29.658471   
min                              -1.791886                   0.213211   
25%                              -0.188029                   1.906355   
50%                               0.207047                   2.775502   
75%                               0.924850                   5.396321   
max                              68.101622                 211.948997   

        Standard deviation of the DM-SNR curve  \
count                              9056.000000   
mean                                 26.313198   
std                                  19.643677   
min                                   7.370432   
25%                                  14.394279   
50%                                  18.299766   
75%                                  28.244943   
max                                 110.642211   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
count                          10000.000000                    9517.000000   
mean                               8.345170                     105.928268   
std                                4.524510                     107.072212   
min                               -2.812353                      -1.976976   
25%                                5.826037                      35.626671   
50%                                8.467912                      83.571656   
75%                               10.736560                     140.455040   
max                               34.539844                    1191.000837   

       target_class  
count  10000.000000  
mean       0.091200  
std        0.287908  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

## Task 1: Implementing Data Quality Plan

### Identifying Issues with Dataset

In [134]:
# Check Data Types of all Columns
df.dtypes

 Mean of the integrated profile                  float64
 Standard deviation of the integrated profile    float64
 Excess kurtosis of the integrated profile       float64
 Skewness of the integrated profile              float64
 Mean of the DM-SNR curve                        float64
 Standard deviation of the DM-SNR curve          float64
 Excess kurtosis of the DM-SNR curve             float64
 Skewness of the DM-SNR curve                    float64
target_class                                     float64
dtype: object

We can see above that all columns have the same datatype of **float64**, so no changes need to be made there. 

In [135]:
# Check for Missing Values
missing = df.isnull().sum(axis=1)
rows_with_missing_values = (missing > 0).sum()
print("Total of: %d rows with missing values" % (rows_with_missing_values))
print("\nTotal of: %d missing values in the dataset" % (missing.sum()))

Total of: 2606 rows with missing values

Total of: 2832 missing values in the dataset


Three columns have quite a few missing values. 

In [136]:
# Check for Duplicates
df[df.duplicated()]

Empty DataFrame
Columns: [ Mean of the integrated profile,  Standard deviation of the integrated profile,  Excess kurtosis of the integrated profile,  Skewness of the integrated profile,  Mean of the DM-SNR curve,  Standard deviation of the DM-SNR curve,  Excess kurtosis of the DM-SNR curve,  Skewness of the DM-SNR curve, target_class]
Index: []

No duplicates present in the dataset.

In [137]:
# Check Column headers
df.columns

Index([' Mean of the integrated profile',
       ' Standard deviation of the integrated profile',
       ' Excess kurtosis of the integrated profile',
       ' Skewness of the integrated profile', ' Mean of the DM-SNR curve',
       ' Standard deviation of the DM-SNR curve',
       ' Excess kurtosis of the DM-SNR curve', ' Skewness of the DM-SNR curve',
       'target_class'],
      dtype='object')

In [138]:
# Check for number of places for each value by each column
def count_decimals(val):
    value_str = str(val)
    if '.' in value_str:
        return len(value_str.split(".")[1])
    else:
        return 0

top_vals = []

for col in df.columns[:8]:
    df['DecimalPlaces'] = df[col].apply(count_decimal_places)
    count = df['DecimalPlaces'].value_counts().head(1)
    top_vals.append(count)

weight = sum(top_count.iloc[0] * top_count.index[0] for top_count in top_vals)
total = sum(top_count.iloc[0] for top_count in top_vals)
avg_dec_places = weight / total

print("Average decimal places rounded to over the feature columns: ", avg_dec_places)

Average decimal places rounded to over the feature columns:  8.396959064327485


In [139]:
#Dropping DecimalPlaces Column
df.drop('DecimalPlaces', axis=1, inplace=True)

After reviewing the dataset visually, I noticed quite a few columns that had values with different counts of decimal places. To identify the average number of decimal places, I first created a function (count_decimals(val)) that turns the value into a string, then checks if there is a decimal point, and if there is it runs the length of the decimal places, otherwise returning 0. I then counted the top value of the count of each column's decimal places. Then by calculating the weight and total, we can find the average decimal places for every column. 

The result of **8.396959064327485** can be rounded down to 8. 

### Correct Issues Identified

#### 1. Testing Imputing of values

In [140]:
df_imputed = df.fillna(df.mean())
df_imputed.describe()

Mean of the integrated profile  \
count                     10000.000000   
mean                        111.103095   
std                          25.640401   
min                           5.812500   
25%                         101.076172   
50%                         115.128906   
75%                         127.015625   
max                         189.734375   

        Standard deviation of the integrated profile  \
count                                   10000.000000   
mean                                       46.509434   
std                                         6.798077   
min                                        24.772042   
25%                                        42.362222   
50%                                        46.933947   
75%                                        50.970903   
max                                        91.808628   

        Excess kurtosis of the integrated profile  \
count                                10000.000000   
mean                                     0.477388   
std                                      0.988930   
min                                     -1.738021   
25%                                      0.057861   
50%                                      0.291774   
75%                                      0.477388   
max                                      8.069522   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
count                         10000.000000               10000.000000   
mean                              1.779757                  12.671227   
std                               6.224056                  29.658471   
min                              -1.791886                   0.213211   
25%                              -0.188029                   1.906355   
50%                               0.207047                   2.775502   
75%                               0.924850                   5.396321   
max                              68.101622                 211.948997   

        Standard deviation of the DM-SNR curve  \
count                             10000.000000   
mean                                 26.313198   
std                                  18.693419   
min                                   7.370432   
25%                                  14.755082   
50%                                  19.563227   
75%                                  26.404585   
max                                 110.642211   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
count                          10000.000000                   10000.000000   
mean                               8.345170                     105.928268   
std                                4.524510                     104.454152   
min                               -2.812353                      -1.976976   
25%                                5.826037                      38.497619   
50%                                8.467912                      88.369044   
75%                               10.736560                     136.473924   
max                               34.539844                    1191.000837   

       target_class  
count  10000.000000  
mean       0.091200  
std        0.287908  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [141]:
df.describe()

Mean of the integrated profile  \
count                     10000.000000   
mean                        111.103095   
std                          25.640401   
min                           5.812500   
25%                         101.076172   
50%                         115.128906   
75%                         127.015625   
max                         189.734375   

        Standard deviation of the integrated profile  \
count                                   10000.000000   
mean                                       46.509434   
std                                         6.798077   
min                                        24.772042   
25%                                        42.362222   
50%                                        46.933947   
75%                                        50.970903   
max                                        91.808628   

        Excess kurtosis of the integrated profile  \
count                                 8595.000000   
mean                                     0.477388   
std                                      1.066709   
min                                     -1.738021   
25%                                      0.025565   
50%                                      0.223402   
75%                                      0.473025   
max                                      8.069522   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
count                         10000.000000               10000.000000   
mean                              1.779757                  12.671227   
std                               6.224056                  29.658471   
min                              -1.791886                   0.213211   
25%                              -0.188029                   1.906355   
50%                               0.207047                   2.775502   
75%                               0.924850                   5.396321   
max                              68.101622                 211.948997   

        Standard deviation of the DM-SNR curve  \
count                              9056.000000   
mean                                 26.313198   
std                                  19.643677   
min                                   7.370432   
25%                                  14.394279   
50%                                  18.299766   
75%                                  28.244943   
max                                 110.642211   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
count                          10000.000000                    9517.000000   
mean                               8.345170                     105.928268   
std                                4.524510                     107.072212   
min                               -2.812353                      -1.976976   
25%                                5.826037                      35.626671   
50%                                8.467912                      83.571656   
75%                               10.736560                     140.455040   
max                               34.539844                    1191.000837   

       target_class  
count  10000.000000  
mean       0.091200  
std        0.287908  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

As we can see above, no statistics(except count) were changed after imputing the values with the mean of each column. This makes sense as it is the mean we are imputing into all of the missing values. Count changed as there are now no missing values!

#### 1. Testing Removing Missing Values

In [142]:
df_no_missing = df.dropna()
df_no_missing.describe()

Mean of the integrated profile  \
count                      7394.000000   
mean                        111.197740   
std                          25.629258   
min                           6.187500   
25%                         101.111328   
50%                         115.179688   
75%                         127.277344   
max                         189.734375   

        Standard deviation of the integrated profile  \
count                                    7394.000000   
mean                                       46.499800   
std                                         6.781050   
min                                        24.772042   
25%                                        42.385134   
50%                                        46.900499   
75%                                        50.984045   
max                                        91.808628   

        Excess kurtosis of the integrated profile  \
count                                 7394.000000   
mean                                     0.477773   
std                                      1.069548   
min                                     -1.738021   
25%                                      0.025952   
50%                                      0.224465   
75%                                      0.473325   
max                                      8.069522   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
count                          7394.000000                7394.000000   
mean                              1.785789                  12.703553   
std                               6.292917                  29.853522   
min                              -1.791886                   0.213211   
25%                              -0.185907                   1.903219   
50%                               0.209895                   2.771321   
75%                               0.932944                   5.430184   
max                              68.101622                 211.948997   

        Standard deviation of the DM-SNR curve  \
count                              7394.000000   
mean                                 26.243085   
std                                  19.528981   
min                                   7.370432   
25%                                  14.358088   
50%                                  18.297041   
75%                                  28.300855   
max                                 110.642211   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
count                           7394.000000                    7394.000000   
mean                               8.360110                     106.374475   
std                                4.546702                     107.458933   
min                               -2.636857                      -1.976976   
25%                                5.816635                      35.515359   
50%                                8.459951                      83.856282   
75%                               10.741820                     140.339170   
max                               34.539844                    1191.000837   

       target_class  
count   7394.000000  
mean       0.090344  
std        0.286693  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

Since we know from our identification cells above that there are 2606 rows with missing values, removing 2606 rows from a 10000 row dataset would remove 26% of the dataset- quite a signifcant amount. In the above cell, I remove the rows with missing values then describe the new dataframe. We can visually observe that after removing rows with missing values the count, mean, standard deviation, min, Q1, median, and Q3 have had values updated. This could possibly lead us to producing inaccurate observations and results if we use this dataset for further analysis.  

### 1. Implementation

In [143]:
df = df.fillna(df.mean())
df.describe()

Mean of the integrated profile  \
count                     10000.000000   
mean                        111.103095   
std                          25.640401   
min                           5.812500   
25%                         101.076172   
50%                         115.128906   
75%                         127.015625   
max                         189.734375   

        Standard deviation of the integrated profile  \
count                                   10000.000000   
mean                                       46.509434   
std                                         6.798077   
min                                        24.772042   
25%                                        42.362222   
50%                                        46.933947   
75%                                        50.970903   
max                                        91.808628   

        Excess kurtosis of the integrated profile  \
count                                10000.000000   
mean                                     0.477388   
std                                      0.988930   
min                                     -1.738021   
25%                                      0.057861   
50%                                      0.291774   
75%                                      0.477388   
max                                      8.069522   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
count                         10000.000000               10000.000000   
mean                              1.779757                  12.671227   
std                               6.224056                  29.658471   
min                              -1.791886                   0.213211   
25%                              -0.188029                   1.906355   
50%                               0.207047                   2.775502   
75%                               0.924850                   5.396321   
max                              68.101622                 211.948997   

        Standard deviation of the DM-SNR curve  \
count                             10000.000000   
mean                                 26.313198   
std                                  18.693419   
min                                   7.370432   
25%                                  14.755082   
50%                                  19.563227   
75%                                  26.404585   
max                                 110.642211   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
count                          10000.000000                   10000.000000   
mean                               8.345170                     105.928268   
std                                4.524510                     104.454152   
min                               -2.812353                      -1.976976   
25%                                5.826037                      38.497619   
50%                                8.467912                      88.369044   
75%                               10.736560                     136.473924   
max                               34.539844                    1191.000837   

       target_class  
count  10000.000000  
mean       0.091200  
std        0.287908  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

Since I identified that by imputing missing values with the mean of each column we would not change any statistics of the dataframe, I decided to impute rather than removing rows with missing values. Count has changed as no missing values are present anymore. 

#### 2. Rounding all columns to 8 Decimal Places

In [144]:
df = df.round(8)
df

Mean of the integrated profile  \
0                          114.859375   
1                           98.406250   
2                          110.359375   
3                          133.960938   
4                          100.109375   
...                               ...   
9995                       122.976562   
9996                       110.257812   
9997                       113.062500   
9998                       123.765625   
9999                       135.523438   

       Standard deviation of the integrated profile  \
0                                         41.452672   
1                                         51.290752   
2                                         42.741544   
3                                         48.455908   
4                                         44.539350   
...                                             ...   
9995                                      48.277108   
9996                                      58.110784   
9997                                      48.387716   
9998                                      41.959706   
9999                                      49.079649   

       Excess kurtosis of the integrated profile  \
0                                       0.188914   
1                                       0.421885   
2                                       0.477388   
3                                      -0.289046   
4                                       0.441861   
...                                          ...   
9995                                   -0.027160   
9996                                    0.132559   
9997                                    0.019552   
9998                                   -0.120701   
9999                                    0.004527   

       Skewness of the integrated profile   Mean of the DM-SNR curve  \
0                                0.591629                   2.579431   
1                               -0.158079                   3.683946   
2                                0.538841                   3.418896   
3                                0.093911                   1.377090   
4                                0.350467                   8.473244   
...                                   ...                        ...   
9995                            -0.080098                   5.120401   
9996                            -0.750325                   2.829431   
9997                            -0.362260                   0.352843   
9998                             0.379044                   7.254181   
9999                            -0.419013                   2.815217   

       Standard deviation of the DM-SNR curve  \
0                                   16.024319   
1                                   23.996595   
2                                   19.090287   
3                                   11.289180   
4                                   32.343502   
...                                       ...   
9995                                28.763490   
9996                                18.674093   
9997                                 8.201329   
9998                                35.665727   
9999                                19.911051   

       Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
0                                 9.183242                     105.904593   
1                                 7.378554                      57.771480   
2                                 7.527202                      67.811225   
3                                13.631908                     105.928268   
4                                 4.532549                      22.332795   
...                                    ...                            ...   
9995                              5.867339                      34.484277   
9996                              9.247513                      98.431192   
9997                             27.385053                     804.482706   
9998

In [145]:
df.describe()

Mean of the integrated profile  \
count                     10000.000000   
mean                        111.103095   
std                          25.640401   
min                           5.812500   
25%                         101.076172   
50%                         115.128906   
75%                         127.015625   
max                         189.734375   

        Standard deviation of the integrated profile  \
count                                   10000.000000   
mean                                       46.509434   
std                                         6.798077   
min                                        24.772042   
25%                                        42.362222   
50%                                        46.933947   
75%                                        50.970903   
max                                        91.808628   

        Excess kurtosis of the integrated profile  \
count                                10000.000000   
mean                                     0.477388   
std                                      0.988930   
min                                     -1.738021   
25%                                      0.057861   
50%                                      0.291774   
75%                                      0.477388   
max                                      8.069522   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
count                         10000.000000               10000.000000   
mean                              1.779757                  12.671227   
std                               6.224056                  29.658471   
min                              -1.791886                   0.213211   
25%                              -0.188029                   1.906355   
50%                               0.207047                   2.775502   
75%                               0.924850                   5.396321   
max                              68.101622                 211.948997   

        Standard deviation of the DM-SNR curve  \
count                             10000.000000   
mean                                 26.313198   
std                                  18.693419   
min                                   7.370432   
25%                                  14.755082   
50%                                  19.563227   
75%                                  26.404585   
max                                 110.642211   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
count                          10000.000000                   10000.000000   
mean                               8.345170                     105.928268   
std                                4.524510                     104.454152   
min                               -2.812353                      -1.976976   
25%                                5.826037                      38.497619   
50%                                8.467912                      88.369044   
75%                               10.736560                     136.473924   
max                               34.539844                    1191.000837   

       target_class  
count  10000.000000  
mean       0.091200  
std        0.287908  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

#### 3. Removing Space from Descriptive Feature Columns

In [146]:
df.columns

Index([' Mean of the integrated profile',
       ' Standard deviation of the integrated profile',
       ' Excess kurtosis of the integrated profile',
       ' Skewness of the integrated profile', ' Mean of the DM-SNR curve',
       ' Standard deviation of the DM-SNR curve',
       ' Excess kurtosis of the DM-SNR curve', ' Skewness of the DM-SNR curve',
       'target_class'],
      dtype='object')

In [154]:
df.columns = df.columns.str.lstrip()
df.columns

Index(['Mean of the integrated profile',
       'Standard deviation of the integrated profile',
       'Excess kurtosis of the integrated profile',
       'Skewness of the integrated profile', 'Mean of the DM-SNR curve',
       'Standard deviation of the DM-SNR curve',
       'Excess kurtosis of the DM-SNR curve', 'Skewness of the DM-SNR curve',
       'target_class'],
      dtype='object')

### Explaination/Justification

        1. I chose to impute the missing values in each column rather than remove the rows with missing values as after testing each method, imputing kept the dataframes statistics (count, min, max, Q1,  Median, and Q3) the same but removing rows changed all but max. This makes sense as it physically removed some data. This could lead to problems and false observations/results down the road in our further analysis as we will be performing operations on data with different statistics. 
        2. After reviewing the dataset manually, I noticed quite a few columns had values with different counts of decimal places. To make all of the data uniform so no potential issues rise later on, I decided to average the amount of decimal places and round all values to that count (of 8). 
        3. During the coding of checking number of decimal places, I noticed an error in the column title/headers- the 8 descriptive feature columns have whitespace before the first word- a potential error for plotting/analysis further down. To fix this I simply removed that whitespace from the first 8 columns. 

## Task 2: Normalization or Standardization

### Analysis

In [155]:
df.describe()

Mean of the integrated profile  \
count                    10000.000000   
mean                       111.103095   
std                         25.640401   
min                          5.812500   
25%                        101.076172   
50%                        115.128906   
75%                        127.015625   
max                        189.734375   

       Standard deviation of the integrated profile  \
count                                  10000.000000   
mean                                      46.509434   
std                                        6.798077   
min                                       24.772042   
25%                                       42.362222   
50%                                       46.933947   
75%                                       50.970903   
max                                       91.808628   

       Excess kurtosis of the integrated profile  \
count                               10000.000000   
mean                                    0.477388   
std                                     0.988930   
min                                    -1.738021   
25%                                     0.057861   
50%                                     0.291774   
75%                                     0.477388   
max                                     8.069522   

       Skewness of the integrated profile  Mean of the DM-SNR curve  \
count                        10000.000000              10000.000000   
mean                             1.779757                 12.671227   
std                              6.224056                 29.658471   
min                             -1.791886                  0.213211   
25%                             -0.188029                  1.906355   
50%                              0.207047                  2.775502   
75%                              0.924850                  5.396321   
max                             68.101622                211.948997   

       Standard deviation of the DM-SNR curve  \
count                            10000.000000   
mean                                26.313198   
std                                 18.693419   
min                                  7.370432   
25%                                 14.755082   
50%                                 19.563227   
75%                                 26.404585   
max                                110.642211   

       Excess kurtosis of the DM-SNR curve  Skewness of the DM-SNR curve  \
count                         10000.000000                  10000.000000   
mean                              8.345170                    105.928268   
std                               4.524510                    104.454152   
min                              -2.812353                     -1.976976   
25%                               5.826037                     38.497619   
50%                               8.467912                     88.369044   
75%                              10.736560                    136.473924   
max                              34.539844                   1191.000837   

       target_class  
count  10000.000000  
mean       0.091200  
std        0.287908  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

### Implementation

### Explaination/Justification

Text

## Task 3: 

## Task 4: 

## Task 5: 

## Task 6:

## Task 7:

## Task 8:

## Task 9: